### PR TITLE
Iox #774 update codecov API and cmake support

### LINKED ISSUE #1031: compile error
## Required information

**Operating system:**
E.g. Ubuntu 18.04 LTS

**Compiler version:**
E.g. GCC 7.4.0

**Observed result or behaviour:**
CMake Error at cmake/IceoryxPackageHelper.cmake:80 (install):
    install TARGETS given target "iceoryx_platform" which does not exist in
    this directory.
Call Stack (most recent call first):
    cmake/IceoryxPackageHelper.cmake:63 (install_target_directories_and_header)
    CMakeLists.txt:226

**Expected result or behaviour:**
build this protect

**Conditions where it occurred / Performed steps:**
non

### PR DESCRIPTION
**Pull request by dkroenke**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes partly #774 
- Closes #1031

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -57,9 +57,12 @@
 - Remove the requirement for INVALID_STATE for the cxx::expected [\#987](https://github.com/eclipse-iceoryx/iceoryx/issues/987)
 - Add unique test identifers [\#988](https://github.com/eclipse-iceoryx/iceoryx/issues/988)
 - Remove `ApplicationPort` and `ApplicationPortData` classes [\#415](https://github.com/eclipse-iceoryx/iceoryx/issues/415)
+- Updating Codecov API and enforce CMake version 3.16 for building iceoryx [\#774](https://github.com/eclipse-iceoryx/iceoryx/issues/774) and [\#1031](https://github.com/eclipse-iceoryx/iceoryx/issues/1031)
 
 **API Breaking Changes:**
 
+The CMake files in iceoryx expect to have CMake version 3.16 or greater installed, otherwise the build fails.
```
Comment (elBoberido) [inline CHANGELOG.md:64]: Maybe add that for ubuntu 18.04 the `cmake-mozilla` package from the `universe` repo could be used

File: doc/website/getting-started/installation.md [hunk 2]
```diff
@@ -7,9 +7,9 @@ All iceoryx libraries are deployed as independent CMake packages. Posh is using
 ### :octicons-package-dependencies-16: Dependencies
 
 - 64-bit hardware (e.g. x86_64 or aarch64; 32-bit hardware might work, but is not supported)
-- [CMake](https://cmake.org), 3.10 or later
+- [CMake](https://cmake.org), 3.16 or later
 - One of the following compilers:
-    - [GCC](https://gcc.gnu.org), 7.4 or later (5.4 currently supported too)
+    - [GCC](https://gcc.gnu.org), 9.3 or later (5.4 currently supported too)
```
Comment (elBoberido) [inline doc/website/getting-started/installation.md:12]: I think for QNX SDP 7.1 we need GCC v8

File: CHANGELOG.md (hunk 1 above)
Comment (dkroenke) [reply CHANGELOG.md:64]: Added this as a hint in Changelog

File: .github/workflows/build-test.yml [hunk 3]
```diff
@@ -23,10 +23,17 @@ jobs:
...
     needs: pre-flight-check
     steps:
       - uses: actions/checkout@v2
+      - uses: egor-tensin/setup-gcc@v1.2
+        with:
+          version: 8
+          platform: x64
+      - uses: jwlawson/actions-setup-cmake@v1.12
+        with:
+          cmake-version: '3.16.3' # version used in Ubuntu 20.04 LTS
```
Comment (dkroenke) [inline .github/workflows/build-test.yml:36]: @elBoberido 
> Would it be possible to have on build with cmake 3.16?

See here, we fix the CMake version to 3.16.3

File: cmake/cyclonedds/CMakeLists.txt [hunk 4]
```diff
@@ -14,7 +14,7 @@
 # limitations under the License.
 #
 # SPDX-License-Identifier: Apache-2.0
-cmake_minimum_required(VERSION 3.5)
```
Comment (elfenpiff) [inline cmake/cyclonedds/CMakeLists.txt:17]: Instead of setting in every file the required cmake version, couldn't we add a generic cmake options file which we include where this is set.
I would bet that the reviewers here check only the files which you touched and not mention the files you may have forgotten to adjust.

File: doc/website/getting-started/installation.md [hunk 5]
```diff
@@ -7,9 +7,9 @@ All iceoryx libraries are deployed as independent CMake packages. Posh is using
 ### :octicons-package-dependencies-16: Dependencies
 
 - 64-bit hardware (e.g. x86_64 or aarch64; 32-bit hardware might work, but is not supported)
-- [CMake](https://cmake.org), 3.10 or later
+- [CMake](https://cmake.org), 3.16 or later
 - One of the following compilers:
-    - [GCC](https://gcc.gnu.org), 7.4 or later (5.4 currently supported too)
+    - [GCC](https://gcc.gnu.org), 8.3 or later (5.4 currently supported too)
```
Comment (elfenpiff) [inline doc/website/getting-started/installation.md:12]: Shouldn't we state that we support gcc >= 5.4? Now it sounds like, 8.3 or greater is supported and by the way also 5.4 but what is with 6 and 7?

File: cmake/cyclonedds/CMakeLists.txt (hunk 4 above)
Comment (elfenpiff) [reply cmake/cyclonedds/CMakeLists.txt:17]: According to this little post this is not possible: https://stackoverflow.com/questions/57638044/can-you-have-a-common-cmake-minimum-included-from-another-file

But maybe some other wise developer knows more stuff?! Or tricks?

Comment (dkroenke) [reply cmake/cyclonedds/CMakeLists.txt:17]: For now it is fine to set them in each file, the effort is not that much when doing a search & replace over the project.
I don't see any special trick for circumventing this, even wrapping this into a custom function does not work since the version is not set in the calling scope.

File: CHANGELOG.md [hunk 6]
```diff
@@ -57,9 +57,13 @@
...
+(Hint: Ubuntu 18 users can install `cmake-mozilla` from the universe PPA)
```
Comment (elBoberido) [inline CHANGELOG.md:65]: It is not a ppa but a repo provided by canonical. It's just off by default

File: doc/website/getting-started/installation.md (hunk 5 above)
Comment (elfenpiff) [reply doc/website/getting-started/installation.md:12]: I understand your argumentation but then I would rephrase it like: 5.4 or later (verified via CI 8.3, ...)
The reason is that there are a lot of versions which we are not testing, see:
8.4, 8.5, 9.0, 9.1, 9.2, 9.3, 9.4, 10.0, 10.1 .... you get the idea, only a minor subset.

And what we do in the CI is completely reasonable, we test the oldest compiler which we are supporting, something in the middle and the newest one.

Comment (dkroenke) [reply doc/website/getting-started/installation.md:12]: Well, this is an installation manual addressing users that are new to iceoryx. Giving them now the range "5.4 or later" will somehow imply that we are fully supporting versions like 6 or 7 too.  Version 5.4 is only an exception for QNX SDP 7.0 that we made. Version 8 is guaranteed to work by CI.

Additionally I would avoid to maintain another list of version numbers of compilers that are tested in the CI because they are subject of change while testing the latest compiler.

File: CHANGELOG.md (hunk 6 above)
Comment (dkroenke) [reply CHANGELOG.md]: fixed it

**Review by elfenpiff — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]

**Review by elBoberido — DISMISSED** [review]

**Review by elBoberido — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]
